### ISSUE TITLE
Performance Issues With VS2017

### ISSUE THREAD

**Issue by cjbhaines** [fsprojects/Paket#2578]
### Description

I spoke to @forki about this who I believe has also spoken to @matthid. When trying to use Paket with VS2017 and the new .NET SDK, build times are incredibly slow, unfortunately to the point where it is unusable.

We have a 60 ish project solution which we tried to convert but had to roll back to VS2015, most of which is .NET Standard and .NET Core (we deploy on Linux). When building the solution in 2015, with the packages restored via a paket restore first, build time was about 20 seconds. This is an enormous project with a lot of code to compile. When we moved to VS2017 the build time went up to over 10 minutes, with VS regularly crashing.

I'm not sure where exactly the issue lies but it seems restore is thrashing on build, even if packages are restored. And it looks like they all wait to lock the packages folder, even though packages are present:

Output from dotnet restore:

Performance:
   - Disk IO: 179 milliseconds
   - Runtime: 6 seconds
  Paket version 5.5.1
  packages folder is locked by paket.exe (PID = 17688). Waiting...
  Performance:
   - Disk IO: 284 milliseconds
   - Runtime: 8 seconds
  Paket version 5.5.1
  packages folder is locked by paket.exe (PID = 16636). Waiting...
  Performance:
   - Disk IO: 157 milliseconds
   - Runtime: 7 seconds

As you can see each project is taking 6-8 seconds * 60 projects = 360 to 480 seconds just waiting to restore, let alone the compile time.

VS setting for parallel project builds: 8

### Repro steps

I have prepared a massive solution for you, which contains 50 projects using the new csproj format. IT consists of:

- .NET 4.6.2 Console App: 10 projects (each referencing 10 .NET Standard 1.5 projects)
- .NET Core 1.1 Console App: 10 projects (each referencing 10 .NET Core Class Library projects)
- .NET Core Class Library: 10 projects (each referencing 10 .NET Standard 1.6 Class Library projects)
- .NET Standard 1.5 Class Library: 10 projects
- .NET Standard 1.6 Class Library: 10 projects

All of the projects reference an arbitrary set of packages from NuGet to simulate a real project. This is fairly random, just a selection from our real project. 

### Expected behavior

Build should still be around 20 seconds, at least less than 30. There is nothing in these projects apart from Program.cs and Class1.cs, compile time should be insignificant. When building for a second time, it should be almost instant as there should be nothing to do.

### Actual behavior

I just ran a dotnet restore and it took nearly 20 minutes to complete and builds basically aren't completing before I give up waiting. 

I am running a Surface Book laptop with 3 GHz i7 Dual Core CPU and 16GB memory.

### Known workarounds

None.

### Next Steps

Would you be able to look at optimising some of the aspects of restore please? Some possible ideas:

- Is it possible to skip restore if you know we are already restored?
- Can we move locking from the entire packages folder to a package specific lock? Which is only take out if we actually need to restore?

I really hope this repro helps, please let me know if I can do anything to assist here.

[VS2017PerfIssue.zip](https://github.com/fsprojects/Paket/files/1191390/VS2017PerfIssue.zip)

**Comment by forki** [comment]
@matthid and myself have already discussed this. 

Following things can be done:
* disable gc from restore (PR already merged) 
* make locked code area smaller
* lazily acquire the lock if we really need to write to packages folder 
* only acquire lock for per package subfolder 
* make second restore a no-op by writing hash files

**Comment by forki** [comment]
I also disc with @matthid that for pure MSBuild 15 solutions we don't need the packages golfer anymore (if we fixup the load scripts). The global nuget folder is enough. That would reduce the amount of work that is to do during restore.

**Comment by matthid** [comment]
related https://github.com/fsprojects/Paket/issues/2141

**Comment by forki** [comment]
we should also make sure the script creation happens only once

**Comment by forki** [comment]
Package extraction currently looks returns the files in a package. This means it scans all directories. We don't need this here.

**Comment by forki** [comment]
@matthid look at https://github.com/fsprojects/Paket/pull/2597 I made GetContent lazy. So restore should not enumerate the files anymore. Yay?

**Comment by forki** [comment]
@cjbhaines we released 5.85 with couple of changes. can you please try it and let me know if it's already improvement?

**Comment by cjbhaines** [comment]
@forki I just gave this a go with 5.85.1. The dotnet resotre from PowerShell outside of VS still took a long time. Paket is using quite a lot of CPU while that happens:

![image](https://user-images.githubusercontent.com/3225747/29231261-742f067e-7ede-11e7-8446-5f808c373557.png)

- dotnet restore on the repro took 12 minutes which is down 8 minutes ish from 5.5.1
- Building inside VS after the external restore took about 3 minutes, which I haven't bothered to wait to complete before
- Building again straight after the first build took about 30 seconds

This is definitely an improvement! Would be good to know why the initial restore still takes so long though. Do you think the invalid profile is causing a problem?

Here is a snapshot from the initial dotnet restore:

Paket version 5.85.1
  The profile 'portable-net45+win8+wp8+wp81+wpa81' is not a known profile. Please tell the package author.
  Performance:
   - Disk IO: 195 milliseconds
   - Runtime: 5 seconds
  Paket version 5.85.1
  The profile 'portable-net45+win8+wp8+wp81+wpa81' is not a known profile. Please tell the package author.
  Performance:
   - Disk IO: 190 milliseconds
   - Runtime: 4 seconds
  Paket version 5.85.1
  The profile 'portable-net45+win8+wp8+wp81+wpa81' is not a known profile. Please tell the package author.
  Performance:
   - Disk IO: 166 milliseconds
   - Runtime: 4 seconds



P.S thanks for finding time to look at this between daddy duties! :)

**Comment by forki** [comment]
Yes one of the issues that I need to figure out is why it even checks
target profiles here. I don't think we need to do that for restore at all.

Am 11.08.2017 11:01 nachm. schrieb "Chris Haines" <notifications@github.com

**Comment by forki** [comment]
ok so it checks the target profiles during lock file parsing. that is expected.

**Comment by cjbhaines** [comment]
Can you print the package ID with that message so we know where it's coming from please?

**Comment by forki** [comment]
I think @matthid was already working on that. But you can find it in the
lockfile

Am 12.08.2017 13:38 schrieb "Chris Haines" <notifications@github.com>:

**Comment by cjbhaines** [comment]
Cool, thanks.

**Comment by matthid** [comment]
I think latest alpha should now be usable as all points of this issue were addressed. Please open a new issue if this still doesn't work for you!

**Comment by forki** [comment]
And please report timings back. Thx

Am 28.08.2017 10:58 vorm. schrieb "Matthias Dittrich" <
notifications@github.com>:

**Comment by cjbhaines** [comment]
This is looking good, I think. Question around explicit restore still though.

**Building**

- Build after running clean is about 10 seconds
- Build second time is about 5 seconds
- Solution rebuild takes about 30 seconds


**dotnet restore**

Something is a little strange that when running dotnet restore from inside VS in the Package Manager Console is that I still see fairly long timings for each project, but this doesn't seem to affect the build time now?:

 Paket version 5.91.0
  Performance:
   - Disk IO: 103 milliseconds
   - Runtime: 3 seconds
  Paket version 5.91.0
  Performance:
   - Disk IO: 120 milliseconds
   - Runtime: 3 seconds
  Paket version 5.91.0
  Performance:
   - Disk IO: 98 milliseconds
   - Runtime: 3 seconds
  Paket version 5.91.0

This takes like 5 minutes to complete still. Calling dotnet restore again immediately after takes just as long.

So it looks like explicit restores are still quite long but we can build quite quickly.

**Comment by matthid** [comment]
5.91.0 doesn't use the optimization.

**Comment by cjbhaines** [comment]
Apologies for not reading that correctly!

Before I go into new timings, it's interesting that I Was now able to build ok inside VS2017 after a restore. The only think I changed between the inital test and just now is that I upgraded to VS2017 15.3, so perhaps that includes some perf fixes of its own in the new SDK?

So, using 5.92.0-beta001 I tried a dotnet restore and it was taking slightly longer per project to restore, about 5 seconds per project. Once it completed I ran dotnet restore again and timings were still about 5 seconds per project.

I then tried an upgrade:

Performance:
 - Resolver: 1 minute, 56 seconds (1 runs)
    - Runtime: 1 minute, 6 seconds
    - Blocked (retrieving package details): 41 seconds (53 times)
    - Blocked (retrieving package versions): 9 seconds (10 times)
    - Not Blocked (retrieving package details): 106 times
    - Not Blocked (retrieving package versions): 149 times
 - Disk IO: 5 seconds
 - Average Download Time: 3 seconds
 - Number of downloads: 15
 - Average Request Time: 665 milliseconds
 - Number of Requests: 606
 - Runtime: 8 minutes, 3 seconds

It seemed to spend the majority of the time doing this:

 - ClassicConsoleApp1/paket.references -> ClassicConsoleApp1/ClassicConsoleApp1.csproj
 - ClassicConsoleApp10/paket.references -> ClassicConsoleApp10/ClassicConsoleApp10.csproj
 - ClassicConsoleApp2/paket.references -> ClassicConsoleApp2/ClassicConsoleApp2.csproj
 - ClassicConsoleApp3/paket.references -> ClassicConsoleApp3/ClassicConsoleApp3.csproj
 - ClassicConsoleApp4/paket.references -> ClassicConsoleApp4/ClassicConsoleApp4.csproj
 - ClassicConsoleApp5/paket.references -> ClassicConsoleApp5/ClassicConsoleApp5.csproj
 - ClassicConsoleApp6/paket.references -> ClassicConsoleApp6/ClassicConsoleApp6.csproj
 - ClassicConsoleApp7/paket.references -> ClassicConsoleApp7/ClassicConsoleApp7.csproj
... truncated

I then ran an install:

Performance:
 - Disk IO: 177 milliseconds
 - Runtime: 4 minutes, 56 seconds

Again, spending all its time listing the same output as above.

But then running dotnet restore, it was lightning fast:

  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary7\NetStandardLibrary7.csproj...
  Restore completed in 377.31 ms for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary4\NetStandardLibrary4.csproj.
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary8\NetStandardLibrary8.csproj...
  Generating MSBuild file C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary5\obj\NetStandardLibrary5.csproj.nuget.g.targets.
  Restore completed in 446.77 ms for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary5\NetStandardLibrary5.csproj.
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary9\NetStandardLibrary9.csproj...
  Generating MSBuild file C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary6\obj\NetStandardLibrary6.csproj.nuget.g.targets.
  Restore completed in 392.83 ms for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary6\NetStandardLibrary6.csproj.
  Generating MSBuild file C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary7\obj\NetStandardLibrary7.csproj.nuget.g.targets.
  Restore completed in 411.86 ms for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary7\NetStandardLibrary7.csproj.
  Generating MSBuild file C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary8\obj\NetStandardLibrary8.csproj.nuget.g.targets.
  Restore completed in 407.88 ms for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary8\NetStandardLibrary8.csproj.
  Generating MSBuild file C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary9\obj\NetStandardLibrary9.csproj.nuget.g.targets.
  Restore completed in 328.92 ms for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary9\NetStandardLibrary9.csproj.


So it looks like there is still huge performance problems with update and install, but restore after running one of those is rapid. Would you like a new issue for update/install?

**Comment by matthid** [comment]
Do you have an hdd or an ssd? All that paket is doing is reading and writing lots of small files... or we still have a problem when you already use an ssd. I'll reopen until I have tested this myself.

It is expected that performance kicks in after install/update and not before.

**Comment by matthid** [comment]
Yes it seems like I indeed killed a bit of performance of install/update. Updating lots of projects historically always was a bit slow, so I'll take a look if we can improve that a bit ...

**Comment by cjbhaines** [comment]
I am running a Surface Book with following spec: 6th Gen Intel Core i7, 512GB SSD, 16GB RAM /dGPU. Paket was hogging about 30% of the CPU during the update/install process.

Thanks for looking further into this.

**Comment by matthid** [comment]
@cjbhaines Yes I could reproduce and yes it was painfully slow on your sample. I had to come up with a bunch of optimizations to get it down to `Runtime: 9 seconds` so I hope paket is now usable again for you.

Obviously I still need to get the PR cleaned up and green and released so it might take a few days until a stable is available.

**Comment by matthid** [comment]
Interestingly it wasn't related to my latest changes at all, it seems to be slow for at least quite some time now....

**Comment by cjbhaines** [comment]
Ok sounds great! You are an absolute hero, the thought of going back to NuGet was soul destroying. Can you drop me a line when you have an alpha available please? Will be more than happy to test for you.

**Comment by matthid** [comment]
Ok please try out the new 5.92.0 release and report back if it works for you

**Comment by cjbhaines** [comment]
Another massive step forward here. Update/install can still take a few minutes so I thought I would print out some timings for different scenarios. Building and rebuiling inside VS seems fairly reliable now.

**Timings for update with deleted packages folder, deleted lock file, deleted paket-files:**

Performance:
 - Resolver: 1 minute, 54 seconds (1 runs)
    - Runtime: 1 minute, 17 seconds
    - Blocked (retrieving package details): 10 seconds (32 times)
    - Blocked (retrieving package versions): 25 seconds (16 times)
    - Not Blocked (retrieving package details): 127 times
    - Not Blocked (retrieving package versions): 143 times
 - Disk IO: 55 seconds
 - Average Download Time: 12 seconds
 - Number of downloads: 1
 - Average Request Time: 1 second
 - Number of Requests: 163
 - Runtime: 2 minutes, 22 seconds

**Update second run:**

Performance:
 - Resolver: 1 minute, 58 seconds (1 runs)
    - Runtime: 1 minute, 31 seconds
    - Blocked (retrieving package details): 6 seconds (29 times)
    - Blocked (retrieving package versions): 19 seconds (14 times)
    - Not Blocked (retrieving package details): 130 times
    - Not Blocked (retrieving package versions): 145 times
 - Disk IO: 540 milliseconds
 - Average Request Time: 1 second
 - Number of Requests: 161
 - Runtime: 2 minutes, 13 seconds

**Install timing after update:**

Performance:
 - Disk IO: 210 milliseconds
 - Runtime: 17 seconds

**Install timing after deleting paket-files:**

Performance:
 - Disk IO: 258 milliseconds
 - Runtime: 18 seconds

**dotnet restore timing after running install:**

PS C:\Dev\Paket\VS2017PerfIssue> dotnet restore
  Paket version 5.92.0
   - C:\Dev\Paket\VS2017PerfIssue\ClassicConsoleApp1\obj\ClassicConsoleApp1.csproj.net462.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\ClassicConsoleApp10\obj\ClassicConsoleApp10.csproj.net462.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\ClassicConsoleApp2\obj\ClassicConsoleApp2.csproj.net462.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\ClassicConsoleApp3\obj\ClassicConsoleApp3.csproj.net462.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\ClassicConsoleApp4\obj\ClassicConsoleApp4.csproj.net462.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\ClassicConsoleApp5\obj\ClassicConsoleApp5.csproj.net462.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\ClassicConsoleApp6\obj\ClassicConsoleApp6.csproj.net462.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\ClassicConsoleApp7\obj\ClassicConsoleApp7.csproj.net462.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\ClassicConsoleApp8\obj\ClassicConsoleApp8.csproj.net462.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\ClassicConsoleApp9\obj\ClassicConsoleApp9.csproj.net462.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\NetCoreClassLibrary1\obj\NetCoreClassLibrary1.csproj.netcoreapp1.1.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\NetCoreClassLibrary10\obj\NetCoreClassLibrary10.csproj.netcoreapp1.1.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\NetCoreClassLibrary2\obj\NetCoreClassLibrary2.csproj.netcoreapp1.1.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\NetCoreClassLibrary3\obj\NetCoreClassLibrary3.csproj.netcoreapp1.1.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\NetCoreClassLibrary4\obj\NetCoreClassLibrary4.csproj.netcoreapp1.1.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\NetCoreClassLibrary5\obj\NetCoreClassLibrary5.csproj.netcoreapp1.1.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\NetCoreClassLibrary6\obj\NetCoreClassLibrary6.csproj.netcoreapp1.1.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\NetCoreClassLibrary7\obj\NetCoreClassLibrary7.csproj.netcoreapp1.1.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\NetCoreClassLibrary8\obj\NetCoreClassLibrary8.csproj.netcoreapp1.1.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\NetCoreClassLibrary9\obj\NetCoreClassLibrary9.csproj.netcoreapp1.1.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\NetCoreConsoleApp1\obj\NetCoreConsoleApp1.csproj.netcoreapp1.1.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\NetCoreConsoleApp10\obj\NetCoreConsoleApp10.csproj.netcoreapp1.1.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\NetCoreConsoleApp2\obj\NetCoreConsoleApp2.csproj.netcoreapp1.1.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\NetCoreConsoleApp3\obj\NetCoreConsoleApp3.csproj.netcoreapp1.1.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\NetCoreConsoleApp4\obj\NetCoreConsoleApp4.csproj.netcoreapp1.1.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\NetCoreConsoleApp5\obj\NetCoreConsoleApp5.csproj.netcoreapp1.1.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\NetCoreConsoleApp6\obj\NetCoreConsoleApp6.csproj.netcoreapp1.1.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\NetCoreConsoleApp7\obj\NetCoreConsoleApp7.csproj.netcoreapp1.1.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\NetCoreConsoleApp8\obj\NetCoreConsoleApp8.csproj.netcoreapp1.1.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\NetCoreConsoleApp9\obj\NetCoreConsoleApp9.csproj.netcoreapp1.1.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary1\obj\NetStandardLibrary1.csproj.netstandard1.5.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary10\obj\NetStandardLibrary10.csproj.netstandard1.5.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary11\obj\NetStandardLibrary11.csproj.netstandard1.6.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary12\obj\NetStandardLibrary12.csproj.netstandard1.6.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary13\obj\NetStandardLibrary13.csproj.netstandard1.6.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary14\obj\NetStandardLibrary14.csproj.netstandard1.6.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary15\obj\NetStandardLibrary15.csproj.netstandard1.6.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary16\obj\NetStandardLibrary16.csproj.netstandard1.6.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary17\obj\NetStandardLibrary17.csproj.netstandard1.6.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary18\obj\NetStandardLibrary18.csproj.netstandard1.6.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary19\obj\NetStandardLibrary19.csproj.netstandard1.6.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary2\obj\NetStandardLibrary2.csproj.netstandard1.5.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary20\obj\NetStandardLibrary20.csproj.netstandard1.6.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary3\obj\NetStandardLibrary3.csproj.netstandard1.5.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary4\obj\NetStandardLibrary4.csproj.netstandard1.5.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary5\obj\NetStandardLibrary5.csproj.netstandard1.5.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary6\obj\NetStandardLibrary6.csproj.netstandard1.5.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary7\obj\NetStandardLibrary7.csproj.netstandard1.5.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary8\obj\NetStandardLibrary8.csproj.netstandard1.5.paket.resolved created
   - C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary9\obj\NetStandardLibrary9.csproj.netstandard1.5.paket.resolved created
  Performance:
   - Disk IO: 127 milliseconds
   - Runtime: 3 minutes
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\ClassicConsoleApp3\ClassicConsoleApp3.csproj...
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\ClassicConsoleApp2\ClassicConsoleApp2.csproj...
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\ClassicConsoleApp1\ClassicConsoleApp1.csproj...
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\ClassicConsoleApp10\ClassicConsoleApp10.csproj...
  Restore completed in 1.93 sec for C:\Dev\Paket\VS2017PerfIssue\ClassicConsoleApp10\ClassicConsoleApp10.csproj.
  Restore completed in 2.01 sec for C:\Dev\Paket\VS2017PerfIssue\ClassicConsoleApp1\ClassicConsoleApp1.csproj.
  Restore completed in 2.01 sec for C:\Dev\Paket\VS2017PerfIssue\ClassicConsoleApp3\ClassicConsoleApp3.csproj.
  Restore completed in 2.05 sec for C:\Dev\Paket\VS2017PerfIssue\ClassicConsoleApp2\ClassicConsoleApp2.csproj.
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\ClassicConsoleApp4\ClassicConsoleApp4.csproj...
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\ClassicConsoleApp5\ClassicConsoleApp5.csproj...
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\ClassicConsoleApp6\ClassicConsoleApp6.csproj...
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\ClassicConsoleApp7\ClassicConsoleApp7.csproj...
  Restore completed in 1.16 sec for C:\Dev\Paket\VS2017PerfIssue\ClassicConsoleApp6\ClassicConsoleApp6.csproj.
  Restore completed in 1.29 sec for C:\Dev\Paket\VS2017PerfIssue\ClassicConsoleApp4\ClassicConsoleApp4.csproj.
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\ClassicConsoleApp8\ClassicConsoleApp8.csproj...
  Restore completed in 1.31 sec for C:\Dev\Paket\VS2017PerfIssue\ClassicConsoleApp5\ClassicConsoleApp5.csproj.
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\ClassicConsoleApp9\ClassicConsoleApp9.csproj...
  Restore completed in 1.45 sec for C:\Dev\Paket\VS2017PerfIssue\ClassicConsoleApp7\ClassicConsoleApp7.csproj.
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetCoreClassLibrary10\NetCoreClassLibrary10.csproj...
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetCoreClassLibrary1\NetCoreClassLibrary1.csproj...
  Restore completed in 1.22 sec for C:\Dev\Paket\VS2017PerfIssue\ClassicConsoleApp8\ClassicConsoleApp8.csproj.
  Restore completed in 1.23 sec for C:\Dev\Paket\VS2017PerfIssue\ClassicConsoleApp9\ClassicConsoleApp9.csproj.
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetCoreClassLibrary2\NetCoreClassLibrary2.csproj...
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetCoreClassLibrary3\NetCoreClassLibrary3.csproj...
  Restore completed in 1.43 sec for C:\Dev\Paket\VS2017PerfIssue\NetCoreClassLibrary10\NetCoreClassLibrary10.csproj.
  Restore completed in 1.45 sec for C:\Dev\Paket\VS2017PerfIssue\NetCoreClassLibrary1\NetCoreClassLibrary1.csproj.
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetCoreClassLibrary4\NetCoreClassLibrary4.csproj...
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetCoreClassLibrary5\NetCoreClassLibrary5.csproj...
  Restore completed in 1.08 sec for C:\Dev\Paket\VS2017PerfIssue\NetCoreClassLibrary2\NetCoreClassLibrary2.csproj.
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetCoreClassLibrary6\NetCoreClassLibrary6.csproj...
  Restore completed in 1.12 sec for C:\Dev\Paket\VS2017PerfIssue\NetCoreClassLibrary3\NetCoreClassLibrary3.csproj.
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetCoreClassLibrary7\NetCoreClassLibrary7.csproj...
  Restore completed in 976.98 ms for C:\Dev\Paket\VS2017PerfIssue\NetCoreClassLibrary4\NetCoreClassLibrary4.csproj.
  Restore completed in 949.21 ms for C:\Dev\Paket\VS2017PerfIssue\NetCoreClassLibrary5\NetCoreClassLibrary5.csproj.
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetCoreClassLibrary8\NetCoreClassLibrary8.csproj...
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetCoreClassLibrary9\NetCoreClassLibrary9.csproj...
  Restore completed in 957.23 ms for C:\Dev\Paket\VS2017PerfIssue\NetCoreClassLibrary6\NetCoreClassLibrary6.csproj.
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetCoreConsoleApp10\NetCoreConsoleApp10.csproj...
  Restore completed in 1.01 sec for C:\Dev\Paket\VS2017PerfIssue\NetCoreClassLibrary7\NetCoreClassLibrary7.csproj.
  Restore completed in 952.24 ms for C:\Dev\Paket\VS2017PerfIssue\NetCoreClassLibrary8\NetCoreClassLibrary8.csproj.
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetCoreConsoleApp1\NetCoreConsoleApp1.csproj...
  Restore completed in 939.38 ms for C:\Dev\Paket\VS2017PerfIssue\NetCoreClassLibrary9\NetCoreClassLibrary9.csproj.
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetCoreConsoleApp2\NetCoreConsoleApp2.csproj...
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetCoreConsoleApp3\NetCoreConsoleApp3.csproj...
  Restore completed in 1.25 sec for C:\Dev\Paket\VS2017PerfIssue\NetCoreConsoleApp10\NetCoreConsoleApp10.csproj.
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetCoreConsoleApp4\NetCoreConsoleApp4.csproj...
  Restore completed in 1.26 sec for C:\Dev\Paket\VS2017PerfIssue\NetCoreConsoleApp1\NetCoreConsoleApp1.csproj.
  Restore completed in 1.23 sec for C:\Dev\Paket\VS2017PerfIssue\NetCoreConsoleApp2\NetCoreConsoleApp2.csproj.
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetCoreConsoleApp5\NetCoreConsoleApp5.csproj...
  Restore completed in 1.26 sec for C:\Dev\Paket\VS2017PerfIssue\NetCoreConsoleApp3\NetCoreConsoleApp3.csproj.
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetCoreConsoleApp6\NetCoreConsoleApp6.csproj...
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetCoreConsoleApp7\NetCoreConsoleApp7.csproj...
  Restore completed in 1.36 sec for C:\Dev\Paket\VS2017PerfIssue\NetCoreConsoleApp4\NetCoreConsoleApp4.csproj.
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetCoreConsoleApp8\NetCoreConsoleApp8.csproj...
  Restore completed in 1.3 sec for C:\Dev\Paket\VS2017PerfIssue\NetCoreConsoleApp5\NetCoreConsoleApp5.csproj.
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetCoreConsoleApp9\NetCoreConsoleApp9.csproj...
  Restore completed in 1.3 sec for C:\Dev\Paket\VS2017PerfIssue\NetCoreConsoleApp6\NetCoreConsoleApp6.csproj.
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary10\NetStandardLibrary10.csproj...
  Restore completed in 1.36 sec for C:\Dev\Paket\VS2017PerfIssue\NetCoreConsoleApp7\NetCoreConsoleApp7.csproj.
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary11\NetStandardLibrary11.csproj...
  Generating MSBuild file C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary10\obj\NetStandardLibrary10.csproj.nuget.g.targets.
  Restore completed in 497.02 ms for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary10\NetStandardLibrary10.csproj.
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary12\NetStandardLibrary12.csproj...
  Generating MSBuild file C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary11\obj\NetStandardLibrary11.csproj.nuget.g.targets.
  Restore completed in 576.2 ms for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary11\NetStandardLibrary11.csproj.
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary13\NetStandardLibrary13.csproj...
  Generating MSBuild file C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary12\obj\NetStandardLibrary12.csproj.nuget.g.targets.
  Restore completed in 425.17 ms for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary12\NetStandardLibrary12.csproj.
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary14\NetStandardLibrary14.csproj...
  Restore completed in 1.38 sec for C:\Dev\Paket\VS2017PerfIssue\NetCoreConsoleApp8\NetCoreConsoleApp8.csproj.
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary15\NetStandardLibrary15.csproj...
  Restore completed in 1.46 sec for C:\Dev\Paket\VS2017PerfIssue\NetCoreConsoleApp9\NetCoreConsoleApp9.csproj.
  Generating MSBuild file C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary13\obj\NetStandardLibrary13.csproj.nuget.g.targets.
  Restore completed in 570.83 ms for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary13\NetStandardLibrary13.csproj.
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary16\NetStandardLibrary16.csproj...
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary17\NetStandardLibrary17.csproj...
  Generating MSBuild file C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary14\obj\NetStandardLibrary14.csproj.nuget.g.targets.
  Restore completed in 619.13 ms for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary14\NetStandardLibrary14.csproj.
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary18\NetStandardLibrary18.csproj...
  Generating MSBuild file C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary15\obj\NetStandardLibrary15.csproj.nuget.g.targets.
  Restore completed in 702.54 ms for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary15\NetStandardLibrary15.csproj.
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary19\NetStandardLibrary19.csproj...
  Generating MSBuild file C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary17\obj\NetStandardLibrary17.csproj.nuget.g.targets.
  Generating MSBuild file C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary16\obj\NetStandardLibrary16.csproj.nuget.g.targets.
  Restore completed in 558.77 ms for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary17\NetStandardLibrary17.csproj.
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary1\NetStandardLibrary1.csproj...
  Restore completed in 647.78 ms for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary16\NetStandardLibrary16.csproj.
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary20\NetStandardLibrary20.csproj...
  Generating MSBuild file C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary18\obj\NetStandardLibrary18.csproj.nuget.g.targets.
  Restore completed in 504.4 ms for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary18\NetStandardLibrary18.csproj.
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary2\NetStandardLibrary2.csproj...
  Generating MSBuild file C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary19\obj\NetStandardLibrary19.csproj.nuget.g.targets.
  Generating MSBuild file C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary1\obj\NetStandardLibrary1.csproj.nuget.g.targets.
  Restore completed in 444.97 ms for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary1\NetStandardLibrary1.csproj.
  Restore completed in 533.17 ms for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary19\NetStandardLibrary19.csproj.
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary3\NetStandardLibrary3.csproj...
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary4\NetStandardLibrary4.csproj...
  Generating MSBuild file C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary20\obj\NetStandardLibrary20.csproj.nuget.g.targets.
  Restore completed in 547.62 ms for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary20\NetStandardLibrary20.csproj.
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary5\NetStandardLibrary5.csproj...
  Generating MSBuild file C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary2\obj\NetStandardLibrary2.csproj.nuget.g.targets.
  Restore completed in 452.97 ms for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary2\NetStandardLibrary2.csproj.
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary6\NetStandardLibrary6.csproj...
  Generating MSBuild file C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary3\obj\NetStandardLibrary3.csproj.nuget.g.targets.
  Generating MSBuild file C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary4\obj\NetStandardLibrary4.csproj.nuget.g.targets.
  Restore completed in 452.25 ms for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary3\NetStandardLibrary3.csproj.
  Restore completed in 428.51 ms for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary4\NetStandardLibrary4.csproj.
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary8\NetStandardLibrary8.csproj...
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary7\NetStandardLibrary7.csproj...
  Generating MSBuild file C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary5\obj\NetStandardLibrary5.csproj.nuget.g.targets.
  Restore completed in 431.55 ms for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary5\NetStandardLibrary5.csproj.
  Restoring packages for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary9\NetStandardLibrary9.csproj...
  Generating MSBuild file C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary6\obj\NetStandardLibrary6.csproj.nuget.g.targets.
  Generating MSBuild file C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary8\obj\NetStandardLibrary8.csproj.nuget.g.targets.
  Restore completed in 561.1 ms for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary6\NetStandardLibrary6.csproj.
  Generating MSBuild file C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary7\obj\NetStandardLibrary7.csproj.nuget.g.targets.
  Restore completed in 347.5 ms for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary8\NetStandardLibrary8.csproj.
  Restore completed in 357.56 ms for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary7\NetStandardLibrary7.csproj.
  Generating MSBuild file C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary9\obj\NetStandardLibrary9.csproj.nuget.g.targets.
  Restore completed in 432.26 ms for C:\Dev\Paket\VS2017PerfIssue\NetStandardLibrary9\NetStandardLibrary9.csproj.

**dotnet restore second run:**

<2 seconds, although dotnet restore doesn't output anything for about 10-15 seconds before Paket starts writing out.

**Comment by jamescrowley** [comment]
@forki @matthid just so I'm clear from our twitter chat, is this the change that went in and was reverted? so this is still an ongoing issue for now? couldn't quite work out if the subsequent 0.92.* releases now up on NuGet included this again or not, but I definitely still have the issue on 0.92.3

**Comment by matthid** [comment]
Yes this is released now.

**Comment by jamescrowley** [comment]
@matthid ok - then I'm still seeing ~6 seconds per project running 0.92.3 in a solution with 100+ projects. Shall I open a new issue or is there additional context I can provide here?

**Comment by matthid** [comment]
Did you `paket install` after updating paket? If yes then sure lets start with a new issue :).

Usually, paket is no longer called in the msbuild process if everything is setup correctly.

**Comment by jamescrowley** [comment]
@matthid solved after a paket install updated the targets file, thanks very much!

**Comment by megafinz** [comment]
Is `paket restore` issue supposed to be fixed?

I'm using 5.120.0. My setup is WPF Desktop application with 52 projects and `auto-restore on`. If I just build my solution, I'm seeing output similar to https://github.com/fsprojects/Paket/issues/2595. If I manually run `paket restore` first (which is quite fast), solution builds very fast then, I don't event see any Paket related stuff in build output.

VSTS build time hopped from 1.5 minutes to 4.5 minutes just by introducing paket auto-restore.

**Comment by forki** [comment]
Yes Auto-restore is not good if you have many projects. It's only solved
for new dotnet sdk projects. I recommend that you go with explicit paket
restore

Am 30.10.2017 16:03 schrieb "Tony" <notifications@github.com>:

**Comment by megafinz** [comment]
Okay, I see, will revert back to explicit package restore step. Thanks.
